chain/ethereum/src/ethereum_adapter.rs

```diff
@@ -414,22 +414,62 @@ impl EthereumAdapter {
         .boxed()
     }
 
+    // Method to determine block_id based on support for EIP-1898
+    fn block_ptr_to_id(&self, block_ptr: &BlockPtr) -> BlockId {
+        // Ganache does not support calls by block hash.
+        // See https://github.com/trufflesuite/ganache-cli/issues/973
+        if !self.supports_eip_1898 {
+            BlockId::Number(block_ptr.number.into())
+        } else {
+            BlockId::Hash(block_ptr.hash_as_h256())
+        }
+    }
+
+    fn code(
+        &self,
+        logger: &Logger,
+        address: Address,
+        block_ptr: BlockPtr,
+    ) -> impl Future<Item = Bytes, Error = EthereumRpcError> + Send {
+        let web3 = self.web3.clone();
+        let logger = Logger::new(&logger, o!("provider" => self.provider.clone()));
+
+        let block_id = self.block_ptr_to_id(&block_ptr);
+        let retry_log_message = format!("eth_getCode RPC call for block {}", block_ptr);
+
+        retry(retry_log_message, &logger)
+            .when(|result| match result {
+                Ok(_) => false,
+                Err(_) => true,
+            })
+            .limit(ENV_VARS.request_retries)
```

Comment:
I think we shouldn't be adding more references to ENV_VARS instead of having those as values in structs. This makes it basically impossible to test and we should be removing these patterns instead of adding to them.